### PR TITLE
[ci-integration] Add flag for error exit code.

### PR DESCRIPTION
**Pull request by elland**
Starts addressing ideas from https://github.com/MangoIV/cabal-audit/issues/27

### CODE REVIEW

File: .envrc [hunk 1]
```diff
@@ -1 +1 @@
-use flake . -Lv
+use flake .#plain-haskell -Lv
```
Comment (MangoIV) [inline .envrc:1]:
```suggestion
use flake . -Lv
```

File: nix/haskell-shell.nix [hunk 2]
```diff
@@ -2,4 +2,5 @@
 hspkgs.shellFor {
   packages = hps: [hps.cabal-audit];
   nativeBuildInputs = [hspkgs.haskell-language-server hspkgs.fourmolu];
+  ignoreCollisions = true;
```
Comment (MangoIV) [inline nix/haskell-shell.nix:5]:
```suggestion
```

File: src/Distribution/Audit.hs [hunk 3]
```diff
@@ -46,7 +46,7 @@ import Security.Advisories (Advisory (..), Keyword (..), ParseAdvisoryError (..)
 import Security.Advisories.Cabal (ElaboratedPackageInfoAdvised, ElaboratedPackageInfoWith (..), matchAdvisoriesForPlan)
 import Security.Advisories.Convert.OSV qualified as OSV
 import Security.Advisories.Filesystem (listAdvisories)
-import System.Exit (exitFailure)
+import System.Exit
```
Comment (MangoIV) [inline src/Distribution/Audit.hs:49]: I would prefer if we could keep full qualified imports as often as possible, if it isn't like 20 imports, I think it's fine

File: src/Distribution/Audit.hs [hunk 4]
```diff
@@ -117,14 +119,24 @@ auditMain = do
...
+          ]
+        liftIO exitFailure
+    `catch` \ex -> do
+      case fromException ex :: Maybe ExitCode of
+        Just (ExitFailure _code) -> exitFailure
+        Just e -> exitWith e
+        Nothing -> pure ()
+      runM $ interpPretty do
+        owo
+          [ ([red, bold], "!!! cabal-audit failed :\n")
```
Comment (MangoIV) [inline src/Distribution/Audit.hs:136]:
```suggestion
          [ ([red, bold], "cabal-audit failed :\n")
```

File: src/Distribution/Audit.hs [hunk 5]
```diff
@@ -117,14 +119,24 @@ auditMain = do
...
-        [ ([red, bold], "cabal-audit failed:\n")
-        , ([red], T.pack $ displayException ex)
-        ]
-      liftIO exitFailure
+      advisories <- buildAdvisories auditConfig nixStyleFlags
+      handleBuiltAdvisories (outputHandle auditConfig) (outputFormat auditConfig) advisories
+      when (auditConfig.failOnWarning && not (null advisories)) $ do
+        owo
+          [ ([red], T.pack (show (length advisories)) <> T.pack " advisories found.")
+          ]
```
Comment (MangoIV) [inline src/Distribution/Audit.hs:127]:
```suggestion
          [([red], T.pack (show (length advisories)) <> T.pack " advisories found.")]
```

**Review by MangoIV — CHANGES_REQUESTED** [review]
looks largely good to me! 

Would be nice if this could be part of a `--ci` flag in future.

File: src/Distribution/Audit.hs (hunk 3 above)
Comment (elland) [reply src/Distribution/Audit.hs]: Good point!

File: src/Distribution/Audit.hs [hunk 6]
```diff
@@ -117,14 +119,23 @@ auditMain = do
...
+      when (auditConfig.failOnWarning && not (null advisories)) $ do
+        owo
+          [([red], T.pack (show (length advisories)) <> T.pack " advisories found.")]
+        liftIO exitFailure
+    `catch` \ex -> do
+      case fromException ex :: Maybe ExitCode of
+        Just (ExitFailure _code) -> exitFailure
+        Just e -> exitWith e
+        Nothing -> pure ()
+      runM $ interpPretty do
```
Comment (MangoIV) [inline src/Distribution/Audit.hs:133]: I think I don't understand this block, why do we not just do the disambiguation on whether to fail or not at the very end underneath the `catch`?

Comment (elland) [reply src/Distribution/Audit.hs]: It doesn't look or read that great and feels a bit redundant. The new version looks better.

**Pull request merged**